File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* p1 = list1;
        ListNode* p2 = list2;
        ListNode head;
        ListNode* prev = &head;


        while (p1 != nullptr && p2 != nullptr) {
            if (p1->val < p2->val) {
                prev->next = p1;
                p1 = p1->next;
            } else {
                prev->next = p2;
                p2 = p2->next;
            }
            prev = prev->next;
        }

        prev->next = (p1 != nullptr) ? p1 : p2;

        return head.next;
    }
};

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        cur1 = list1
        cur2 = list2
        head = ListNode()
        res = head

        while cur1 is not None or cur2 is not None:
            if cur2 is None:
                res.next = ListNode(cur1.val)
                cur1 = cur1.next
            elif cur1 is None:
                res.next = ListNode(cur2.val)
                cur2 = cur2.next
            elif cur1.val <= cur2.val:
                res.next = ListNode(cur1.val)
                cur1 = cur1.next
            else:
                res.next = ListNode(cur2.val)
                cur2 = cur2.next

            res = res.next
        
        return head.next
